dist/common/scripts/scylla_util.py

```diff
@@ -105,6 +109,141 @@ def curl(url, byte=False):
             if (retries >= max_retries):
                 raise
 
+class gcp_instance:
+    """Describe several aspects of the current GCP instance"""
+
+    from psutil import virtual_memory
+    from psutil import cpu_count
+
+    def __init__(self):
+        self.__type = None
+        self.__cpu = None
+        self.__memoryGB = None
+        self.__nvmeDiskCount = None
+
+    @staticmethod
+    def is_gce_instance():
+      """Check if it's GCE instance via DNS lookup to metadata server."""
+      import socket
+      try:
+        addrlist = socket.getaddrinfo('metadata.google.internal', 80)
+      except socket.gaierror:
+        return False
+      for res in addrlist:
+        af, socktype, proto, canonname, sa = res
+        if (af == socket.AF_INET):
+          addr, port = sa
+          if (addr == "169.254.169.254"):
+            return True
+      return False
+
+    def __instance_metadata(self, path):
+        """query GCP metadata server, recursively!"""
+        return curl("http://169.254.169.254/computeMetadata/v1/instance/" + path+"?recursive=true", headers="Metadata-Flavor: Google")
+        #169.254.169.254 is metadata.google.internal
+
+    def _non_root_nvmes(self):
+        """get list of nvme disks from os, filter away if one of them is root"""
+        nvme_re = re.compile(r"nvme\d+n\d+$")
+
+        root_dev_candidates = [x for x in psutil.disk_partitions() if x.mountpoint == "/"]
+        if len(root_dev_candidates) != 1:
+            raise Exception("found more than one disk mounted at root'".format(root_dev_candidates))
+
+        root_dev = root_dev_candidates[0].device
+        if root_dev.startswith("/dev/mapper"):
+            raise Exception("mapper used for root, not checking if nvme is used".format(root_dev))
+
+        nvmes_present = list(filter(nvme_re.match, os.listdir("/dev")))
+        return {"root": [root_dev], "ephemeral": [x for x in nvmes_present if not root_dev.startswith(os.path.join("/dev/", x))]}
```

Comment:
How can you be sure it is ephemeral? In AWS there can be nvme EBS drives.

Reply:
this is legacy code similar to AWS, I can use other name for the list
the idea was to make some calls between AWS and GCP generic and share them between the classes, but I guess it's not a good idea

Reply:
Making the code generic is a good idea. But what about my question?

Reply:
you cannot be sure, that's why the count from metadata is the primary source
that said in GCP you cannot have nvme attached disks at the moment
+ my tests that tested other SSD drives attached showed their naming is standard and doesn't use /dev/nvme*
so GCP uses /dev/nvme only for local attached ssd disks

Reply:
"at the moment" may change in the next moment. But our images will remain.

So I'm trying to make it robust against changes.

Reply:
so you're right Avi,
and this is sample metadata:
"disks":[{"deviceName":"scylla-disk-test2","index":0,"interface":"SCSI","mode":"READ_WRITE","type":"PERSISTENT-SSD"},{"deviceName":"local-ssd-0","index":1,"interface":"NVME","mode":"READ_WRITE","type":"LOCAL-SSD"},{"deviceName":"local-ssd-1","index":2,"interface":"NVME","mode":"READ_WRITE","type":"LOCAL-SSD"}]

none of above is visible in OS though (since the data visible to OS is just partitioned 3TB card)
(so mapping it to OS is N/A)

I have a bad feeling we won't be able to be robust here unless Google fixes their device exposure to VM and add a way to map drives in some sane way
from above I really could just see the count of disks (which however is enough for us to set the limit)
what can break here is that we could pick wrong disk for size check and in worst case the system won't be set directly, but iotune will run for it (fwiw this sounds robust enough for me)

and we can of course add a smarter check once Google really adds support for persistent nvme disks
- fwiw AWS also doesn't solve this problem - so how about we address this once this is fully clear and address it for all clouds we support?
(I will file a tracking issue for it)

Reply:
Let me understand - you assigned 8 375GB disks, and it shows up as 2 disks in the metadata?

Let's at least check the instance type. So if a newer instance comes along with different disk setup, we don't assume anything about it.

Reply:
no no, 2 disks in metadata show as 2 nvme disks in OS, but it's not possible to create the mapping between metadata and what info OS provides (e.g. local-ssd-1 from metadata is not to be found anywhere in OS and similar for other metadata about local nvmes)

I can add back the instance check, it was dropped due to ratio checks are more generic ... but then I think it's not needed, the code will be safe enough (and same approach is used in AWS without any error reports)

Reply:
It is not safe. The numbers in the table are for those instance types and not others. 

If tomorrow GCP adds an instance type with different local disks, we don't want to use those numbers.

Reply:
ok, so then it makes sense to check if we have 32 vcpus and then use the numbers for 16 or 24 disks
otherwise run iotune
and also respect https://cloud.google.com/compute/docs/disks#local_ssds_and_machine_types and restrict based on them
should we also use the flat numbers directly from the table instead of our measurements?

Reply:
also if we want to be diligent, we shouldn't even proceed with machines that don't support local ssds as per https://cloud.google.com/compute/docs/machine-types#memory-optimized_machine_type_family

Reply:
I don't understand why we need to count vcpus. Can't we determine the instance type from the metadata?

We can support machines without local disks, but need to run iotune (on the persistent disk). Unless we can determine the persistent disk's performance.

Reply:
We can use the table's numbers directly, if our kernel is able to reach them.

Reply:
per tests with iotune without Pavels fixes, we can reach more iops now, but I have no problem to lower it to what Google says it supports (note that I won't do other geos testing or retest with new iotune, since in scope of this it doesn't make sense - I am taking Google numbers as granted)

yes, instance type is fetched from metadata, I just dropped the checks, but I will add them back as per above docs

and yes, if machine won't fall into the blessed category iotune will be run(this is already the behaviour)